### PR TITLE
Ema

### PR DESCRIPTION
**Pull request by jshilong**
I implement  EMA algorithm as a Hook, it support resume checkpoint when train process terminated  by accident.
This algorithm can stable the train process and allow user to use bigger learning  rate to get higher performance.

### CODE REVIEW

File: mmcv/runner/hooks/ema.py [hunk 1]
```diff
@@ -0,0 +1,88 @@
...
+        warm_up (int): during first warm_up steps, we may use smaller momentum
+            to update ema parameters more slowly.
+        resume_from (str): the checkpoint path
+    """
+
+    def __init__(self,
+                 momentum=0.1,
+                 interval=1,
+                 warm_up=100,
+                 resume_from=None):
```
[the remark this answers is not in the document]
Comment (jshilong) [reply mmcv/runner/hooks/ema.py:32]: I use `resume_from` to keep same name with original config

[the remark this answers is not in the document]
Comment (ZwwWayne) [reply mmcv/runner/hooks/ema.py:33]: > I use `resume_from` to keep same name with original config

Can we change that? I also think `checkpoint` is better than `resume_from`.

File: mmcv/runner/hooks/ema.py [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+            self.param_ema_buffer[name] = buffer_name
+            model.register_buffer(buffer_name, value.data.clone())
+        self.model_buffers = dict(model.named_buffers(recurse=True))
+        if self.resume_from is not None:
+            runner.resume(self.resume_from)
+
+    def after_train_iter(self, runner):
+        """Update ema parameter every self.interval iterations."""
+        curr_step = runner.iter
+        # We warm up the momentum considering the instabilitygit  at beginning
```
Comment (ZwwWayne) [inline mmcv/runner/hooks/ema.py:62]: instabilitygit -> instability

File: mmcv/runner/base_runner.py [hunk 3]
```diff
@@ -291,10 +291,13 @@ def resume(self,
...
-            checkpoint = self.load_checkpoint(
-                checkpoint,
-                map_location=lambda storage, loc: storage.cuda(device_id))
+            if torch.cuda.is_available():
+                device_id = torch.cuda.current_device()
+                checkpoint = self.load_checkpoint(
+                    checkpoint,
+                    map_location=lambda storage, loc: storage.cuda(device_id))
+            else:
+                checkpoint = self.load_checkpoint(checkpoint)
```
Comment (ZwwWayne) [inline mmcv/runner/base_runner.py:300]: if cuda is not available here, we should map weights to cpu?

Comment (jshilong) [reply mmcv/runner/base_runner.py:300]: `self.load_checkpoint` has default value

File: mmcv/runner/hooks/ema.py [hunk 4]
```diff
@@ -0,0 +1,88 @@
+from ...parallel import is_module_wrapper
+from ..hooks.hook import HOOKS, Hook
+
+
+@HOOKS.register_module()
+class EmaHook(Hook):
+    r"""Exponential Moving Average Hook
+
+    Use Exponential Moving Average on all parameters of model in training
+    process.all parameter has a ema backup, which update by the formula
```
Comment (ZwwWayne) [inline mmcv/runner/hooks/ema.py:10]: '.all parameter has' -> '. All parameters have'

File: mmcv/runner/hooks/ema.py [hunk 5]
```diff
@@ -0,0 +1,88 @@
...
+            \text{Xema_{t}} +  \text{momentum} \times X_t
+
+    Args:
+        momentum (float): The momentum used for update ema parameter.
+            Defaults to 0.9998.
+        interval (int): Update ema parameter every interval iteration.
+            Defaults to 1.
+        warm_up (int): During first warm_up steps, we may use smaller momentum
+            to update ema parameters more slowly. Defaults to 100.
+        checkpoint (str): The checkpoint path. Defaults to None.
```
Comment (hellock) [inline mmcv/runner/hooks/ema.py:25]: Maybe renamed to `resume_from`?

File: mmcv/runner/hooks/ema.py [hunk 6]
```diff
@@ -0,0 +1,88 @@
...
+        """We load parameter values from ema backup to model before the
+        EvalHook."""
+        self.swap_ema_parameters()
+
+    def before_train_epoch(self, runner):
+        """We recover model's parameter from ema backup after last epoch's the
+        EvalHook."""
+        self.swap_ema_parameters()
+
+    def swap_ema_parameters(self):
```
Comment (hellock) [inline mmcv/runner/hooks/ema.py:82]: `_swap_ema_parameters`

File: mmcv/runner/hooks/ema.py [hunk 7]
```diff
@@ -0,0 +1,88 @@
...
+    process. All parameters have a ema backup, which update by the formula
+    as below. EMAHook takes priority over EvalHook and CheckpointSaverHook.
+
+        .. math::
+
+            \text{Xema_{t+1}} = (1 - \text{momentum}) \times
+            \text{Xema_{t}} +  \text{momentum} \times X_t
+
+    Args:
+        momentum (float): The momentum used for update ema parameter.
```
Comment (hellock) [inline mmcv/runner/hooks/ema.py:19]: update -> updating

**Comment by www516717402** [comment]
`self.model_parameters` param not update by after_train_iter ?

**Pull request merged**